sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessorBase.cs

```diff
@@ -0,0 +1,729 @@
+    {
+        /// <summary>The seed to use for initializing random number generated for a given thread-specific instance.</summary>
+        private static int s_randomSeed = Environment.TickCount;
+
+        /// <summary>The random number generator to use for a specific thread.</summary>
+        private static readonly ThreadLocal<Random> RandomNumberGenerator = new ThreadLocal<Random>(() => new Random(Interlocked.Increment(ref s_randomSeed)), false);
+
+        /// <summary>The primitive for synchronizing access during start and close operations.</summary>
+        private readonly SemaphoreSlim RunningTaskSemaphore = new SemaphoreSlim(1, 1);
+
+        /// <summary>
+        ///   The fully qualified Event Hubs namespace that the processor is associated with.  This is likely
+        ///   to be similar to <c>{yournamespace}.servicebus.windows.net</c>.
+        /// </summary>
+        ///
+        public abstract string FullyQualifiedNamespace { get; }
+
+        /// <summary>
+        ///   The name of the Event Hub that the processor is connected to, specific to the
+        ///   Event Hubs namespace that contains it.
+        /// </summary>
+        ///
+        public abstract string EventHubName { get; }
+
+        /// <summary>
+        ///   The name of the consumer group this event processor is associated with.  Events will be
+        ///   read only in the context of this group.
+        /// </summary>
+        ///
+        public abstract string ConsumerGroup { get; }
+
+        /// <summary>
+        ///   A unique name used to identify this event processor.
+        /// </summary>
+        ///
+        public abstract string Identifier { get; }
+
+        /// <summary>
+        ///   The active policy which governs retry attempts for the
+        ///   processor.
+        /// </summary>
+        ///
+        protected abstract EventHubsRetryPolicy RetryPolicy { get; }
+
+        /// <summary>
+        ///   The minimum amount of time to be elapsed between two load balancing verifications.
+        /// </summary>
+        ///
+        protected virtual TimeSpan LoadBalanceUpdate => TimeSpan.FromSeconds(10);
+
+        /// <summary>
+        ///   The minimum amount of time for an ownership to be considered expired without further updates.
+        /// </summary>
+        ///
+        protected virtual TimeSpan OwnershipExpiration => TimeSpan.FromSeconds(30);
+
+        /// <summary>
+        ///   A <see cref="CancellationTokenSource"/> instance to signal the request to cancel the current running task.
+        /// </summary>
+        ///
+        private CancellationTokenSource RunningTaskTokenSource { get; set; }
+
+        /// <summary>
+        ///   The set of currently active partition processing tasks issued by this event processor.  Partition ids are
+        ///   used as keys.
+        /// </summary>
+        ///
+        protected ConcurrentDictionary<string, Task> ActivePartitionProcessors { get; private set; }
```

Comment:
Any reason why we don't create one instance and just clear it on stop?

Reply:
Should we store the cancellation and the task together?    In a tuple or little nested struct?

Reply:
I wanted to stop holding the reference while the processor is stopped. I think it would make sense to do that once `CloseAsync` is called, but that's something that only exists in the derived class.

Reply:
I think using a tuple might be a good idea. We could use it to "force" (not really, but kinda) the user to give us a way of signaling the task should be stopped.

Reply:
- Why the dictionary being `protected` would make it impossible to use the tuple? I'm ok with having a struct, anyway, but I'll leave it to post-GA.
- I think your point is reasonable and we should just clear dicts instead creating new ones every time. I'll get this updated in this PR.